File: intelitor-v5.2/lib/cepaf/src/Cepaf/Observability/Fractal/ContentRouter.fs
namespace Cepaf.Observability.Fractal

open System
open System.Collections.Concurrent

/// Content Router for intelligent log backend selection.
/// Routes logs to appropriate storage based on fractal level, content type, and retention.
/// STAMP Compliance: SC-LOG-010 (L1/L2 ephemeral, L4/L5 persistent)
module ContentRouter =

    // ============================================================
    // TYPES
    // ============================================================

    /// Backend storage types
    [<RequireQualifiedAccess>]
    type Backend =
        /// In-memory ring buffer (L1 ephemeral)
        | Memory

        /// Local WAL for durability (L2-L3)
        | WAL

        /// TimescaleDB for time-series (L3-L4)
        | TimescaleDB

        /// PostgreSQL for structured data (L4)
        | PostgreSQL

        /// S3-compatible object storage (L5 archival)
        | ObjectStore

        /// OpenTelemetry Collector (all levels)
        | OTLP

        /// Console output (debug)
        | Console

        /// Custom backend handler
        | Custom of string

    /// Retention policy
    type RetentionPolicy = {
        /// Minimum retention duration
        MinRetention: TimeSpan

        /// Maximum retention duration
        MaxRetention: TimeSpan

        /// Whether to archive before deletion
        ArchiveOnExpiry: bool

        /// Compression level for storage
        CompressionLevel: int
    }

    /// Routing rule
    type RoutingRule = {
        /// Rule identifier
        Id: string

        /// Key expression pattern to match
        KeyExpr: string

        /// Compiled pattern
        CompiledExpr: KeyExpression.CompiledExpr option

        /// Minimum fractal level
        MinLevel: FractalLevel

        /// Maximum fractal level
        MaxLevel: FractalLevel

        /// Target backends
        Backends: Backend list

        /// Retention policy
        Retention: RetentionPolicy

        /// Priority (higher = evaluated first)
        Priority: int

        /// Whether rule is enabled
        Enabled: bool
    }

    /// Routing decision
    type RoutingDecision = {
        /// Selected backends
        Backends: Backend list

        /// Applicable retention policy
        Retention: RetentionPolicy

        /// Rule that matched
        MatchedRule: string option

        /// Whether to emit (based on filtering)
        ShouldEmit: bool
    }

    /// Router state
    type RouterState = {
        /// Routing rules
        Rules: ConcurrentDictionary<string, RoutingRule>

        /// Default backends by level
        DefaultBackends: ConcurrentDictionary<FractalLevel, Backend list>

        /// Default retention by level
        DefaultRetention: ConcurrentDictionary<FractalLevel, RetentionPolicy>

        /// Backend availability
        BackendHealth: ConcurrentDictionary<Backend, bool>

        /// Statistics
        mutable RouteCount: int64
        mutable FallbackCount: int64
    }

    // ============================================================
    // DEFAULT CONFIGURATIONS
    // ============================================================

    /// Default retention policies by level
    let private defaultRetentionByLevel =
        [
            (FractalLevel.L1, { MinRetention = TimeSpan.FromMinutes(5.0); MaxRetention = TimeSpan.FromHours(1.0); ArchiveOnExpiry = false; CompressionLevel = 0 })
            (FractalLevel.L2, { MinRetention = TimeSpan.FromHours(1.0); MaxRetention = TimeSpan.FromDays(1.0); ArchiveOnExpiry = false; CompressionLevel = 1 })
            (FractalLevel.L3, { MinRetention = TimeSpan.FromDays(7.0); MaxRetention = TimeSpan.FromDays(30.0); ArchiveOnExpiry = true; CompressionLevel = 6 })
            (FractalLevel.L4, { MinRetention = TimeSpan.FromDays(30.0); MaxRetention = TimeSpan.FromDays(365.0); ArchiveOnExpiry = true; CompressionLevel = 9 })
            (FractalLevel.L5, { MinRetention = TimeSpan.FromDays(365.0); MaxRetention = TimeSpan.FromDays(3650.0); ArchiveOnExpiry = true; CompressionLevel = 9 })
        ]
        |> Map.ofList

    /// Default backends by level
    let private defaultBackendsByLevel =
        [
            (FractalLevel.L1, [Backend.Memory; Backend.OTLP])
            (FractalLevel.L2, [Backend.WAL; Backend.OTLP])
            (FractalLevel.L3, [Backend.TimescaleDB; Backend.OTLP])
            (FractalLevel.L4, [Backend.PostgreSQL; Backend.TimescaleDB; Backend.OTLP])
            (FractalLevel.L5, [Backend.PostgreSQL; Backend.ObjectStore; Backend.OTLP])
        ]
        |> Map.ofList

    // ============================================================
    // STATE MANAGEMENT
    // ============================================================

    let mutable private state: RouterState option = None
    let private initLock = obj()

    /// Initialize the router
    let initialize () =
        lock initLock (fun () ->
            let defaultBackends = ConcurrentDictionary<FractalLevel, Backend list>()
            let defaultRetention = ConcurrentDictionary<FractalLevel, RetentionPolicy>()
            let backendHealth = ConcurrentDictionary<Backend, bool>()

            // Set defaults
            for (level, backends) in defaultBackendsByLevel |> Map.toSeq do
                defaultBackends.[level] <- backends

            for (level, retention) in defaultRetentionByLevel |> Map.toSeq do
                defaultRetention.[level] <- retention

            // All backends healthy by default
            [Backend.Memory; Backend.WAL; Backend.TimescaleDB; Backend.PostgreSQL; Backend.ObjectStore; Backend.OTLP; Backend.Console]
            |> List.iter (fun b -> backendHealth.[b] <- true)

            state <- Some {
                Rules = ConcurrentDictionary<string, RoutingRule>()
                DefaultBackends = defaultBackends
                DefaultRetention = defaultRetention
                BackendHealth = backendHealth
                RouteCount = 0L
                FallbackCount = 0L
            }
        )

    /// Get or initialize state
    let private getState () =
        match state with
        | Some s -> s
        | None ->
            initialize()
            state.Value

    // ============================================================
    // RULE MANAGEMENT
    // ============================================================

    /// Add a routing rule
    let addRule (rule: RoutingRule) : Result<unit, string> =
        let s = getState()

        // Compile key expression
        let compiled =
            match KeyExpression.compile rule.KeyExpr with
            | Ok c -> Some c
            | Error _ -> None

        let ruleWithCompiled = { rule with CompiledExpr = compiled }
        s.Rules.[rule.Id] <- ruleWithCompiled
        Ok ()

    /// Remove a routing rule
    let removeRule (ruleId: string) : bool =
        let s = getState()
        let mutable removed = Unchecked.defaultof<RoutingRule>
        s.Rules.TryRemove(ruleId, &removed)

    /// Enable/disable a rule
    let setRuleEnabled (ruleId: string) (enabled: bool) : bool =
        let s = getState()
        match s.Rules.TryGetValue(ruleId) with
        | true, rule ->
            s.Rules.[ruleId] <- { rule with Enabled = enabled }
            true
        | false, _ -> false

    /// Get all rules
    let getRules () : RoutingRule list =
        let s = getState()
        s.Rules.Values |> Seq.toList

    // ============================================================
    // BACKEND HEALTH
    // ============================================================

    /// Set backend health status
    let setBackendHealth (backend: Backend) (healthy: bool) =
        let s = getState()
        s.BackendHealth.[backend] <- healthy

    /// Check if backend is healthy
    let isBackendHealthy (backend: Backend) : bool =
        let s = getState()
        match s.BackendHealth.TryGetValue(backend) with
        | true, health -> health
        | false, _ -> false

    /// Get all healthy backends
    let getHealthyBackends () : Backend list =
        let s = getState()
        s.BackendHealth
        |> Seq.filter (fun kvp -> kvp.Value)
        |> Seq.map (fun kvp -> kvp.Key)
        |> Seq.toList

    // ============================================================
    // ROUTING LOGIC
    // ============================================================

    /// Find matching rule for an entry
    let private findMatchingRule (s: RouterState) (entry: FractalLogEntry) : RoutingRule option =
        s.Rules.Values
        |> Seq.filter (fun r -> r.Enabled)
        |> Seq.filter (fun r ->
            // Level check
            FractalLevel.toInt entry.FractalLevel >= FractalLevel.toInt r.MinLevel &&
            FractalLevel.toInt entry.FractalLevel <= FractalLevel.toInt r.MaxLevel
        )
        |> Seq.filter (fun r ->
            // Key expression check
            match r.CompiledExpr with
            | Some compiled -> KeyExpression.matches compiled entry.Key
            | None -> entry.Key.StartsWith(r.KeyExpr) || r.KeyExpr = "**"
        )
        |> Seq.sortByDescending (fun r -> r.Priority)
        |> Seq.tryHead

    /// Filter backends by health
    let private filterHealthyBackends (s: RouterState) (backends: Backend list) : Backend list =
        backends
        |> List.filter (fun b ->
            match s.BackendHealth.TryGetValue(b) with
            | true, health -> health
            | false, _ -> true  // Assume healthy if unknown
        )

    /// Route a log entry to appropriate backends
    let route (entry: FractalLogEntry) : RoutingDecision =
        let s = getState()
        System.Threading.Interlocked.Increment(&s.RouteCount) |> ignore

        // Find matching rule
        match findMatchingRule s entry with
        | Some rule ->
            let healthyBackends = filterHealthyBackends s rule.Backends

            // Fallback if no healthy backends
            let finalBackends =
                if healthyBackends.IsEmpty then
                    System.Threading.Interlocked.Increment(&s.FallbackCount) |> ignore
                    [Backend.Console]  // Always available
                else
                    healthyBackends

            {
                Backends = finalBackends
                Retention = rule.Retention
                MatchedRule = Some rule.Id
                ShouldEmit = true
            }

        | None ->
            // Use defaults based on level
            let defaultBackends =
                match s.DefaultBackends.TryGetValue(entry.FractalLevel) with
                | true, backends -> backends
                | false, _ -> [Backend.OTLP]

            let healthyBackends = filterHealthyBackends s defaultBackends

            let retention =
                match s.DefaultRetention.TryGetValue(entry.FractalLevel) with
                | true, ret -> ret
                | false, _ -> defaultRetentionByLevel.[FractalLevel.L3]

            {
                Backends = if healthyBackends.IsEmpty then [Backend.Console] else healthyBackends
                Retention = retention
                MatchedRule = None
                ShouldEmit = true
            }

    /// Batch route multiple entries
    let routeBatch (entries: FractalLogEntry list) : (FractalLogEntry * RoutingDecision) list =
        entries |> List.map (fun e -> (e, route e))

    // ============================================================
    // BACKEND ADAPTERS
    // ============================================================

    /// Backend adapter interface
    type IBackendAdapter =
        abstract member Write: FractalLogEntry -> Async<Result<unit, string>>
        abstract member WriteBatch: FractalLogEntry list -> Async<Result<int, string>>
        abstract member IsHealthy: unit -> bool
        abstract member Name: string

    /// Registry for backend adapters
    let private adapters = ConcurrentDictionary<Backend, IBackendAdapter>()

    /// Register a backend adapter
    let registerAdapter (backend: Backend) (adapter: IBackendAdapter) =
        adapters.[backend] <- adapter

    /// Get adapter for backend
    let getAdapter (backend: Backend) : IBackendAdapter option =
        match adapters.TryGetValue(backend) with
        | true, adapter -> Some adapter
        | false, _ -> None

    /// Write entry to all routed backends
    let writeToBackends (entry: FractalLogEntry) : Async<Result<unit, string>> =
        async {
            let decision = route entry

            if not decision.ShouldEmit then
                return Ok ()
            else

            let! results =
                decision.Backends
                |> List.choose (fun b -> getAdapter b |> Option.map (fun a -> (b, a)))
                |> List.map (fun (_, adapter) ->
                    async {
                        try
                            return! adapter.Write entry
                        with ex ->
                            return Error ex.Message
                    }
                )
                |> Async.Parallel

            let errors = results |> Array.choose (function Error e -> Some e | Ok _ -> None)

            if errors.Length = 0 then
                return Ok ()
            elif errors.Length < results.Length then
                // Partial success
                return Ok ()
            else
                return Error (String.Join("; ", errors))
        }

    // ============================================================
    // PREDEFINED RULES
    // ============================================================

    /// Create rule for security audit logs
    let securityAuditRule () : RoutingRule =
        {
            Id = "security-audit"
            KeyExpr = "Indrajaal/Security/**"
            CompiledExpr = None
            MinLevel = FractalLevel.L3
            MaxLevel = FractalLevel.L5
            Backends = [Backend.PostgreSQL; Backend.ObjectStore; Backend.OTLP]
            Retention = {
                MinRetention = TimeSpan.FromDays(365.0)
                MaxRetention = TimeSpan.FromDays(3650.0)  // 10 years
                ArchiveOnExpiry = true
                CompressionLevel = 9
            }
            Priority = 100
            Enabled = true
        }

    /// Create rule for cognitive/AI logs
    let cognitiveRule () : RoutingRule =
        {
            Id = "cognitive"
            KeyExpr = "Indrajaal/Cortex/**"
            CompiledExpr = None
            MinLevel = FractalLevel.L4
            MaxLevel = FractalLevel.L5
            Backends = [Backend.PostgreSQL; Backend.TimescaleDB; Backend.ObjectStore]
            Retention = {
                MinRetention = TimeSpan.FromDays(90.0)
                MaxRetention = TimeSpan.FromDays(730.0)  // 2 years
                ArchiveOnExpiry = true
                CompressionLevel = 9
            }
            Priority = 90
            Enabled = true
        }

    /// Create rule for alarm processing
    let alarmRule () : RoutingRule =
        {
            Id = "alarms"
            KeyExpr = "Indrajaal/Alarms/**"
            CompiledExpr = None
            MinLevel = FractalLevel.L3
            MaxLevel = FractalLevel.L5
            Backends = [Backend.TimescaleDB; Backend.PostgreSQL; Backend.OTLP]
            Retention = {
                MinRetention = TimeSpan.FromDays(30.0)
                MaxRetention = TimeSpan.FromDays(365.0)
                ArchiveOnExpiry = true
                CompressionLevel = 6
            }
            Priority = 80
            Enabled = true
        }

    /// Create rule for debug/ephemeral logs
    let debugRule () : RoutingRule =
        {
            Id = "debug"
            KeyExpr = "**"
            CompiledExpr = None
            MinLevel = FractalLevel.L1
            MaxLevel = FractalLevel.L2
            Backends = [Backend.Memory; Backend.Console]
            Retention = {
                MinRetention = TimeSpan.FromMinutes(5.0)
                MaxRetention = TimeSpan.FromHours(1.0)
                ArchiveOnExpiry = false
                CompressionLevel = 0
            }
            Priority = 1  // Lowest priority, catch-all
            Enabled = true
        }

    /// Initialize with default rules
    let initializeWithDefaults () =
        initialize()
        addRule (securityAuditRule()) |> ignore
        addRule (cognitiveRule()) |> ignore
        addRule (alarmRule()) |> ignore
        addRule (debugRule()) |> ignore

    // ============================================================
    // STATISTICS & DIAGNOSTICS
    // ============================================================

    /// Get router statistics
    let getStats () =
        let s = getState()
        {|
            RouteCount = s.RouteCount
            FallbackCount = s.FallbackCount
            RuleCount = s.Rules.Count
            EnabledRules = s.Rules.Values |> Seq.filter (fun r -> r.Enabled) |> Seq.length
            HealthyBackends = getHealthyBackends() |> List.length
            TotalBackends = s.BackendHealth.Count
        |}

    /// Reset statistics
    let resetStats () =
        let s = getState()
        s.RouteCount <- 0L
        s.FallbackCount <- 0L

    /// Reset to uninitialized state
    let reset () =
        lock initLock (fun () ->
            state <- None
        )

File: intelitor-v5.2/lib/indrajaal/observability/fractal/content_router.ex
defmodule Indrajaal.Observability.Fractal.ContentRouter do
  @moduledoc """
  Content Router for intelligent log backend selection.

  Routes logs to appropriate storage based on fractal level, content type, and retention.
  Supports multi-cast to multiple backends with Zenoh-style key expression matching.

  ## STAMP Compliance

  - SC-LOG-001: Async dispatch (never block)
  - SC-LOG-006: L3+ logs MUST use HLC timestamps
  - SC-LOG-010: L1/L2 ephemeral, L4/L5 persistent

  ## Default Routing Rules

  - `Indrajaal/**/L5` -> SIEM + SigNoz (dual write)
  - `Indrajaal/Security/**` -> SIEM
  - `Indrajaal/**/error` -> ErrorTracker
  - Default -> SigNoz only

  ## Performance

  - Target: < 1us per route decision
  - O(1) for exact matches via ETS lookup
  - O(n) for wildcard matching (n = number of rules)
  """

  use GenServer
  require Logger

  alias Indrajaal.Observability.Fractal.KeyExpression

  # ============================================================
  # TYPES
  # ============================================================

  @type backend ::
          :memory
          | :wal
          | :timescale_db
          | :postgresql
          | :object_store
          | :otlp
          | :signoz
          | :siem
          | :error_tracker
          | :console
          | :blockchain_ledger
          | :zenoh
          | {:custom, String.t()}

  @type retention_policy :: %{
          min_retention: non_neg_integer(),
          max_retention: non_neg_integer(),
          archive_on_expiry: boolean(),
          compression_level: non_neg_integer()
        }

  @type routing_rule :: %{
          id: String.t(),
          key_expr: String.t(),
          compiled_expr: map() | nil,
          min_level: atom(),
          max_level: atom(),
          backends: [backend()],
          retention: retention_policy(),
          priority: integer(),
          enabled: boolean()
        }

  @type routing_decision :: %{
          backends: [backend()],
          retention: retention_policy(),
          matched_rule: String.t() | nil,
          should_emit: boolean()
        }

  # ============================================================
  # CONSTANTS
  # ============================================================

  @level_to_int %{l1: 1, l2: 2, l3: 3, l4: 4, l5: 5}

  @default_retention_by_level %{
    l1: %{
      min_retention: 5 * 60_000,
      max_retention: 60 * 60_000,
      archive_on_expiry: false,
      compression_level: 0
    },
    l2: %{
      min_retention: 60 * 60_000,
      max_retention: 24 * 60 * 60_000,
      archive_on_expiry: false,
      compression_level: 1
    },
    l3: %{
      min_retention: 7 * 24 * 60 * 60_000,
      max_retention: 30 * 24 * 60 * 60_000,
      archive_on_expiry: true,
      compression_level: 6
    },
    l4: %{
      min_retention: 30 * 24 * 60 * 60_000,
      max_retention: 365 * 24 * 60 * 60_000,
      archive_on_expiry: true,
      compression_level: 9
    },
    l5: %{
      min_retention: 365 * 24 * 60 * 60_000,
      max_retention: 3650 * 24 * 60 * 60_000,
      archive_on_expiry: true,
      compression_level: 9
    }
  }

  @default_backends_by_level %{
    # SC-ZENOH-INT-001: All levels route to Zenoh for real-time streaming
    l1: [:memory, :otlp, :zenoh],
    l2: [:wal, :otlp, :zenoh],
    l3: [:timescale_db, :otlp, :zenoh],
    l4: [:postgresql, :timescale_db, :otlp, :zenoh],
    l5: [:postgresql, :object_store, :otlp, :zenoh]
  }

  # ============================================================
  # CLIENT API
  # ============================================================

  @doc """
  Start the ContentRouter GenServer.
  """
  @spec start_link(keyword()) :: GenServer.on_start()
  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
  end

  @doc """
  Route a log entry to appropriate backends.

  This is the HOT PATH - must complete in < 1us for optimal performance.
  Uses ETS for O(1) lookups where possible.

  ## Parameters

  - `entry` - Fractal log entry map with :key, :level fields

  ## Returns

  - `routing_decision()` with selected backends and retention policy
  """
  @spec route(map()) :: routing_decision()
  def route(entry) do
    ensure_started()
    do_route(entry)
  end

  @doc """
  Route multiple entries in a batch (more efficient for bulk operations).
  """
  @spec route_batch([map()]) :: [{map(), routing_decision()}]
  def route_batch(entries) when is_list(entries) do
    Enum.map(entries, fn entry -> {entry, route(entry)} end)
  end

  @doc """
  Add a new routing rule.
  """
  @spec add_rule(routing_rule()) :: :ok | {:error, term()}
  def add_rule(rule) do
    ensure_started()
    GenServer.call(__MODULE__, {:add_rule, rule})
  end

  @doc """
  Remove a routing rule by ID.
  """
  @spec remove_rule(String.t()) :: :ok | {:error, :not_found}
  def remove_rule(rule_id) do
    ensure_started()
    GenServer.call(__MODULE__, {:remove_rule, rule_id})
  end

  @doc """
  Enable or disable a rule.
  """
  @spec set_rule_enabled(String.t(), boolean()) :: :ok | {:error, :not_found}
  def set_rule_enabled(rule_id, enabled) do
    ensure_started()
    GenServer.call(__MODULE__, {:set_rule_enabled, rule_id, enabled})
  end

  @doc """
  Get all routing rules.
  """
  @spec get_rules() :: [routing_rule()]
  def get_rules do
    ensure_started()

    :fractal_routing_rules
    |> :ets.tab2list()
    |> Enum.map(fn {_id, rule} -> rule end)
  end

  @doc """
  Set backend health status.
  """
  @spec set_backend_health(backend(), boolean()) :: :ok
  def set_backend_health(backend, healthy) do
    ensure_started()
    :ets.insert(:fractal_backend_health, {backend, healthy})
    :ok
  end

  @doc """
  Check if a backend is healthy.
  """
  @spec backend_healthy?(backend()) :: boolean()
  def backend_healthy?(backend) do
    ensure_started()

    case :ets.lookup(:fractal_backend_health, backend) do
      [{^backend, healthy}] -> healthy
      # Assume healthy if unknown
      [] -> true
    end
  end

  @doc """
  Get all healthy backends.
  """
  @spec healthy_backends() :: [backend()]
  def healthy_backends do
    ensure_started()

    :fractal_backend_health
    |> :ets.tab2list()
    |> Enum.filter(fn {_b, healthy} -> healthy end)
    |> Enum.map(fn {b, _} -> b end)
  end

  @doc """
  Get router statistics.
  """
  @spec get_stats() :: map()
  def get_stats do
    ensure_started()
    GenServer.call(__MODULE__, :get_stats)
  end

  @doc """
  Reset statistics counters.
  """
  @spec reset_stats() :: :ok
  def reset_stats do
    ensure_started()
    GenServer.cast(__MODULE__, :reset_stats)
  end

  @doc """
  Initialize with default routing rules.
  """
  @spec initialize_with_defaults() :: :ok
  def initialize_with_defaults do
    ensure_started()

    # Security audit rule - highest priority
    add_rule(security_audit_rule())

    # L5 cognitive logs to SIEM + SigNoz (dual write)
    add_rule(l5_dual_write_rule())

    # Error tracking rule
    add_rule(error_tracking_rule())

    # Alarms rule
    add_rule(alarms_rule())

    # Cognitive/AI rule
    add_rule(cognitive_rule())

    # BlockchainLedger rule for immutable L5 audit trails
    add_rule(blockchain_audit_rule())

    # Debug/ephemeral rule (lowest priority catch-all)
    add_rule(debug_rule())

    :ok
  end

  @doc """
  Create rule for blockchain-backed immutable audit logs.
  """
  @spec blockchain_audit_rule() :: routing_rule()
  def blockchain_audit_rule do
    %{
      id: "blockchain-audit",
      key_expr: "Indrajaal/Audit/**",
      compiled_expr: nil,
      min_level: :l5,
      max_level: :l5,
      backends: [:blockchain_ledger, :postgresql, :object_store],
      retention: %{
        # 10 years
        min_retention: 3650 * 24 * 60 * 60_000,
        # 20 years
        max_retention: 7300 * 24 * 60 * 60_000,
        archive_on_expiry: true,
        compression_level: 9
      },
      # High priority after security-audit
      priority: 98,
      enabled: true
    }
  end

  # ============================================================
  # PREDEFINED RULES
  # ============================================================

  @doc """
  Create rule for security audit logs.
  """
  @spec security_audit_rule() :: routing_rule()
  def security_audit_rule do
    %{
      id: "security-audit",
      key_expr: "Indrajaal/Security/**",
      compiled_expr: nil,
      min_level: :l3,
      max_level: :l5,
      backends: [:siem, :postgresql, :object_store, :otlp],
      retention: %{
        min_retention: 365 * 24 * 60 * 60_000,
        max_retention: 3650 * 24 * 60 * 60_000,
        archive_on_expiry: true,
        compression_level: 9
      },
      priority: 100,
      enabled: true
    }
  end

  @doc """
  Create rule for L5 logs requiring dual write to SIEM and SigNoz.
  """
  @spec l5_dual_write_rule() :: routing_rule()
  def l5_dual_write_rule do
    %{
      id: "l5-dual-write",
      key_expr: "Indrajaal/**/L5",
      compiled_expr: nil,
      min_level: :l5,
      max_level: :l5,
      backends: [:siem, :signoz, :postgresql, :object_store],
      retention: %{
        min_retention: 365 * 24 * 60 * 60_000,
        max_retention: 3650 * 24 * 60 * 60_000,
        archive_on_expiry: true,
        compression_level: 9
      },
      priority: 95,
      enabled: true
    }
  end

  @doc """
  Create rule for error tracking.
  """
  @spec error_tracking_rule() :: routing_rule()
  def error_tracking_rule do
    %{
      id: "error-tracking",
      key_expr: "Indrajaal/**/error",
      compiled_expr: nil,
      min_level: :l1,
      max_level: :l5,
      backends: [:error_tracker, :postgresql, :otlp],
      retention: %{
        min_retention: 30 * 24 * 60 * 60_000,
        max_retention: 365 * 24 * 60 * 60_000,
        archive_on_expiry: true,
        compression_level: 6
      },
      priority: 90,
      enabled: true
    }
  end

  @doc """
  Create rule for alarm processing.
  """
  @spec alarms_rule() :: routing_rule()
  def alarms_rule do
    %{
      id: "alarms",
      key_expr: "Indrajaal/Alarms/**",
      compiled_expr: nil,
      min_level: :l3,
      max_level: :l5,
      backends: [:timescale_db, :postgresql, :otlp],
      retention: %{
        min_retention: 30 * 24 * 60 * 60_000,
        max_retention: 365 * 24 * 60 * 60_000,
        archive_on_expiry: true,
        compression_level: 6
      },
      priority: 80,
      enabled: true
    }
  end

  @doc """
  Create rule for cognitive/AI logs.
  """
  @spec cognitive_rule() :: routing_rule()
  def cognitive_rule do
    %{
      id: "cognitive",
      key_expr: "Indrajaal/Cortex/**",
      compiled_expr: nil,
      min_level: :l4,
      max_level: :l5,
      backends: [:postgresql, :timescale_db, :object_store],
      retention: %{
        min_retention: 90 * 24 * 60 * 60_000,
        max_retention: 730 * 24 * 60 * 60_000,
        archive_on_expiry: true,
        compression_level: 9
      },
      priority: 85,
      enabled: true
    }
  end

  @doc """
  Create rule for debug/ephemeral logs.
  """
  @spec debug_rule() :: routing_rule()
  def debug_rule do
    %{
      id: "debug",
      key_expr: "**",
      compiled_expr: nil,
      min_level: :l1,
      max_level: :l2,
      backends: [:memory, :console],
      retention: %{
        min_retention: 5 * 60_000,
        max_retention: 60 * 60_000,
        archive_on_expiry: false,
        compression_level: 0
      },
      priority: 1,
      enabled: true
    }
  end

  # ============================================================
  # GENSERVER CALLBACKS
  # ============================================================

  @impl true
  def init(_opts) do
    # Create ETS tables for fast lookups
    create_ets_tables()

    # Initialize backend health
    initialize_backend_health()

    state = %{
      route_count: 0,
      fallback_count: 0,
      started_at: DateTime.utc_now()
    }

    Logger.info("[ContentRouter] Initialized successfully")
    {:ok, state}
  end

  @impl true
  def handle_call({:add_rule, rule}, _from, state) do
    # Compile key expression
    compiled =
      case KeyExpression.compile(rule.key_expr) do
        {:ok, expr} -> expr
        {:error, _} -> nil
      end

    rule_with_compiled = Map.put(rule, :compiled_expr, compiled)
    :ets.insert(:fractal_routing_rules, {rule.id, rule_with_compiled})

    # Emit telemetry
    :telemetry.execute(
      [:fractal, :router, :rule_added],
      %{count: 1},
      %{rule_id: rule.id, key_expr: rule.key_expr}
    )

    {:reply, :ok, state}
  end

  @impl true
  def handle_call({:remove_rule, rule_id}, _from, state) do
    case :ets.lookup(:fractal_routing_rules, rule_id) do
      [{^rule_id, _}] ->
        :ets.delete(:fractal_routing_rules, rule_id)
        {:reply, :ok, state}

      [] ->
        {:reply, {:error, :not_found}, state}
    end
  end

  @impl true
  def handle_call({:set_rule_enabled, rule_id, enabled}, _from, state) do
    case :ets.lookup(:fractal_routing_rules, rule_id) do
      [{^rule_id, rule}] ->
        updated = Map.put(rule, :enabled, enabled)
        :ets.insert(:fractal_routing_rules, {rule_id, updated})
        {:reply, :ok, state}

      [] ->
        {:reply, {:error, :not_found}, state}
    end
  end

  @impl true
  def handle_call(:get_stats, _from, state) do
    rules = :ets.tab2list(:fractal_routing_rules)
    enabled_rules = Enum.count(rules, fn {_id, r} -> r.enabled end)
    healthy = healthy_backends()

    stats = %{
      route_count: state.route_count,
      fallback_count: state.fallback_count,
      rule_count: length(rules),
      enabled_rules: enabled_rules,
      healthy_backends: length(healthy),
      total_backends: :ets.info(:fractal_backend_health, :size),
      uptime_seconds: DateTime.diff(DateTime.utc_now(), state.started_at)
    }

    {:reply, stats, state}
  end

  @impl true
  def handle_cast(:reset_stats, state) do
    {:noreply, %{state | route_count: 0, fallback_count: 0}}
  end

  @impl true
  def handle_cast({:update_stats, :fallback}, state) do
    {:noreply, %{state | fallback_count: state.fallback_count + 1}}
  end

  @impl true
  def handle_cast({:update_stats, :route}, state) do
    {:noreply, %{state | route_count: state.route_count + 1}}
  end

  @impl true
  def handle_info(:cleanup_expired, state) do
    # Periodic cleanup of stale data if needed
    {:noreply, state}
  end

  # ============================================================
  # PRIVATE: ROUTING LOGIC
  # ============================================================

  defp do_route(entry) do
    key = Map.get(entry, :key, "")
    level = Map.get(entry, :level, :l4)

    # Find matching rule
    case find_matching_rule(key, level) do
      nil ->
        # Use defaults based on level
        default_routing(level)

      rule ->
        # Filter backends by health
        healthy_backends = filter_healthy_backends(rule.backends)

        # Fallback if no healthy backends
        final_backends =
          if Enum.empty?(healthy_backends) do
            update_fallback_count()
            [:console]
          else
            healthy_backends
          end

        %{
          backends: final_backends,
          retention: rule.retention,
          matched_rule: rule.id,
          should_emit: true
        }
    end
  end

  defp find_matching_rule(key, level) do
    level_int = level_to_int(level)

    :fractal_routing_rules
    |> :ets.tab2list()
    |> Enum.filter(fn {_id, rule} ->
      rule.enabled and
        level_int >= level_to_int(rule.min_level) and
        level_int <= level_to_int(rule.max_level) and
        matches_key_expr?(rule, key)
    end)
    |> Enum.sort_by(fn {_id, rule} -> -rule.priority end)
    |> List.first()
    |> case do
      nil -> nil
      {_id, rule} -> rule
    end
  end

  defp matches_key_expr?(rule, key) do
    case rule.compiled_expr do
      nil ->
        # Fallback to simple matching
        String.starts_with?(key, rule.key_expr) or rule.key_expr == "**"

      compiled ->
        KeyExpression.matches?(compiled, key)
    end
  end

  defp filter_healthy_backends(backends) do
    Enum.filter(backends, fn backend ->
      case :ets.lookup(:fractal_backend_health, backend) do
        [{^backend, healthy}] -> healthy
        [] -> true
      end
    end)
  end

  defp default_routing(level) do
    default_backends = Map.get(@default_backends_by_level, level, [:otlp])
    healthy_backends = filter_healthy_backends(default_backends)

    retention = Map.get(@default_retention_by_level, level, @default_retention_by_level[:l3])

    %{
      backends: if(Enum.empty?(healthy_backends), do: [:console], else: healthy_backends),
      retention: retention,
      matched_rule: nil,
      should_emit: true
    }
  end

  defp level_to_int(level), do: Map.get(@level_to_int, level, 4)

  defp update_fallback_count do
    # Update via GenServer to maintain state
    GenServer.cast(__MODULE__, {:update_stats, :fallback})
  end

  # ============================================================
  # PRIVATE: ETS INITIALIZATION
  # ============================================================

  defp create_ets_tables do
    # Rules table
    case :ets.whereis(:fractal_routing_rules) do
      :undefined ->
        :ets.new(:fractal_routing_rules, [:named_table, :set, :public, read_concurrency: true])

      _ ->
        :ok
    end

    # Backend health table
    case :ets.whereis(:fractal_backend_health) do
      :undefined ->
        :ets.new(:fractal_backend_health, [:named_table, :set, :public, read_concurrency: true])

      _ ->
        :ok
    end
  end

  defp initialize_backend_health do
    # All supported backends including BlockchainLedger for immutable L5 audit trails
    # and Zenoh for real-time streaming to F# cockpit (SC-ZENOH-INT-001)
    backends = [
      :memory,
      :wal,
      :timescale_db,
      :postgresql,
      :object_store,
      :otlp,
      :signoz,
      :siem,
      :error_tracker,
      :console,
      :blockchain_ledger,
      :zenoh
    ]

    Enum.each(backends, fn backend ->
      :ets.insert(:fractal_backend_health, {backend, true})
    end)
  end

  defp ensure_started do
    case Process.whereis(__MODULE__) do
      nil ->
        # Start if not running (for testing or lazy initialization)
        case start_link([]) do
          {:ok, _pid} -> :ok
          {:error, {:already_started, _pid}} -> :ok
          error -> error
        end

      _pid ->
        :ok
    end
  end
end
